File: src/main/java/io/github/mdsimmo/bomberman/utils/Box.java
package io.github.mdsimmo.bomberman.utils;

import org.bukkit.Chunk;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.entity.Entity;
import org.bukkit.util.Vector;

import java.util.ArrayList;
import java.util.List;

public final class Box {

	public final Dim loc;
	public final Dim size;
	public final World world;

	public Box(World world, Dim loc, Dim size) {
		this.world = world;
		this.loc = loc;
		this.size = size;
	}

	public Box(World world, int x, int y, int z, int xSize, int ySize, int zSize) {
		this(world, new Dim(x, y, z), new Dim(xSize, ySize, zSize));
	}

	public Box (Location l, Vector size) {
		this(l, size.getBlockX(), size.getBlockY(), size.getBlockZ());
	}

	public Box(Location l, int xSize, int ySize, int zSize) {
		this(l.getWorld(), l.getBlockX(), l.getBlockY(), l.getBlockZ(), xSize, ySize, zSize);
	}
	
	public Box(Location min, Location max) {
		this(min,
				max.getBlockX() - min.getBlockX(),
				max.getBlockY() - min.getBlockY(),
				max.getBlockZ() - min.getBlockZ());
	}

	public Box(Location min, Dim size) {
		this(min.getWorld(), new Dim(min.getBlockX(), min.getBlockY(), min.getBlockZ()), size);
	}


	public boolean contains( Location l ) {
		return contains( l.getWorld(), l.getBlockX(), l.getBlockY(), l.getBlockZ() );
	}
	
	public boolean contains( World world, int x, int y, int z ) {
		if ( !world.equals( this.world ) )
			return false;
		return x >= loc.x && x < loc.x + size.x
			&& y >= loc.y && y < loc.y + size.y
			&& z >= loc.z && z < loc.z + size.z;
	}
	
	public Location corner() {
		return new Location(world, loc.x, loc.y, loc.z);
	}
	
	public Location fromCorner(int xAdd, int yAdd, int zAdd) {
		return new Location(world, loc.x + xAdd, loc.y + yAdd, loc.z + zAdd);
	}
	
	public Location fromCorner(Vector v) {
		return new Location(world, loc.x + v.getX(), loc.y + v.getY(), loc.z + v.getZ());
	}

	public List<Entity> getEntities() {
		List<Entity> entities = new ArrayList<>();
		// the "+ 16" is to make sure the chunks at the edge are also included
		for (int i = loc.x; i < (loc.x + size.x + 16); i += 16) {
			for (int k = loc.z; k < (loc.z + size.z + 16); k += 16) {
				Chunk chunk = world.getBlockAt(i, 1, k).getChunk();
				for (Entity entity : chunk.getEntities()) {
					if (contains(entity.getLocation()))
						entities.add(entity);
				}
			}
		}
		return entities;
	}

	@Override
	public String toString() {
		return "corner at: (" + loc + "); size: " + size;
	}

	@Override
	public int hashCode() {
		int hash = 31;
		hash = hash * 31 + world.hashCode();
		hash = hash * 31 + loc.hashCode();
		hash = hash * 31 + size.hashCode();
		return hash;
	}


	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Box) {
			Box other = (Box) obj;
			return other.world.equals(world) && other.loc.equals(loc) && other.size.equals(size);
		} else {
			return false;
		}
	}
}

File: src/main/java/io/github/mdsimmo/bomberman/utils/Box.kt
package io.github.mdsimmo.bomberman.utils

import org.bukkit.Location
import org.bukkit.World
import org.bukkit.entity.Entity
import java.util.*
import java.util.stream.Stream
import javax.annotation.CheckReturnValue
import kotlin.streams.asStream

@CheckReturnValue
class Box(val world: World, val p1: Dim, val p2: Dim) {

    val size: Dim get() = Dim(p2.x - p1.x + 1, p2.y - p1.y + 1, p2.z - p1.z + 1)

    constructor(world: World, x: Int, y: Int, z: Int, xSize: Int, ySize: Int, zSize: Int) : this(world, Dim(x, y, z), Dim(x+xSize-1, y+ySize-1, z+zSize-1)) {}

    constructor(l: Location, xSize: Int, ySize: Int, zSize: Int) : this(l.world!!, l.blockX, l.blockY, l.blockZ, xSize, ySize, zSize) {}

    operator fun contains(l: Location): Boolean {
        return contains(l.world, l.blockX, l.blockY, l.blockZ)
    }

    fun contains(world: World?, x: Int, y: Int, z: Int): Boolean {
        return if (world != this.world) false else x >= p1.x && x <= p1.x + size.x && y >= p1.y && y <= p1.y + size.y && z >= p1.z && z <= p1.z + size.z
    }

    val entities: List<Entity>
        get() {
            val entities: MutableList<Entity> = ArrayList()
            // the "+ 16" is to make sure the chunks at the edge are also included
            var i = p1.x
            while (i < p1.x + size.x + 16) {
                var k = p1.z
                while (k < p1.z + size.z + 16) {
                    val chunk = world.getBlockAt(i, 1, k).chunk
                    for (entity in chunk.entities) {
                        if (contains(entity.location)) entities.add(entity)
                    }
                    k += 16
                }
                i += 16
            }
            return entities
        }

    override fun toString(): String {
        return "Box $p1 - $p2 : (${size.volume()} blocks)"
    }

    override fun hashCode(): Int {
        var hash = 31
        hash = hash * 31 + world.hashCode()
        hash = hash * 31 + p1.hashCode()
        hash = hash * 31 + p2.hashCode()
        return hash
    }

    fun stream(): Stream<Location> = sequence {
        for (x in p1.x..p2.x) {
            for (y in p1.y..p2.y) {
                for (z in p1.z..p2.z) {
                    yield(Location(world, x.toDouble(), y.toDouble(), z.toDouble()))
                }
            }
        }
    }.asStream()

    override fun equals(other: Any?): Boolean {
        return if (other is Box) {
            other.world == world && other.p1 == p1 && other.p2 == p2
        } else {
            false
        }
    }

}